query/src/main/java/com/connexta/replication/query/QueryService.java

```diff
@@ -0,0 +1,219 @@
+      FilterIndex filterIndex = queryServiceTools.getOrCreateFilterIndex(filter);
+
+      // holds the modified time of each task put in the queue for this filter
+      // this will be used to grab the latest modified time after
+      Set<Instant> modifiedTimes = new HashSet<>();
+
+      try {
+        getNewMetadataForFilter(filter, filterIndex)
+            .map(metadatum -> metadataToTask(metadatum, filter.getPriority()))
+            .forEach(task -> queueTaskAndPutLastModifiedInSet(task, modifiedTimes));
+      } catch (AdapterException e) {
+        LOGGER.debug(
+            "Couldn't query site {}. Stopping queries for this site until next polling attempt.",
+            site.getName(),
+            e);
+        return;
+      } catch (QueueException qe) {
+        LOGGER.debug(
+            "Failed to put a task in queue for site {}. Stopping queuing of tasks for this site. Will resume at next polling attempt.",
+            site.getName(),
+            qe);
+        return;
+      }
+      saveLastModifiedInFilterIndex(filterIndex, modifiedTimes);
+    }
+  }
+
+  private Stream<Metadata> getNewMetadataForFilter(Filter filter, FilterIndex filterIndex)
+      throws AdapterException {
+    // get last modified date
+    final Date indexDate = filterIndex.getModifiedSince().map(Date::from).orElse(null);
+    if (indexDate == null) {
+      LOGGER.trace("No index for filter {} found.", filter.getName());
+    }
+
+    // not implementing failure retries here
+    QueryRequest queryRequest =
+        new QueryRequestImpl(
+            filter.getFilter(),
+            Collections.emptyList(), // ignoring excluded nodes (origins tag)
+            Collections.emptyList(), // ignoring failed items
+            indexDate);
+
+    Iterable<Metadata> changeSet = adapter.query(queryRequest).getMetadata();
+    return StreamSupport.stream(changeSet.spliterator(), false);
+  }
+
+  private TaskInfo metadataToTask(Metadata metadata, byte filterPriority) {
+    TaskInfoImpl task = new TaskInfoImpl();
+    task.setPriority(filterPriority);
+    task.setOperation(determineOperationType());
+
+    if (metadata.getResourceSize() > 0) { // determine if there is a resource
+      final ResourceInfoImpl resourceInfo = new ResourceInfoImpl(metadata);
+      task.setResourceInfo(resourceInfo);
+    }
+
+    final MetadataInfo metadataInfo = new DdfMetadataInfoImpl("Metacard XML", metadata);
+    task.setMetadatas(Stream.of(metadataInfo));
+
+    task.setLastModified(metadata.getMetadataModified().toInstant());
+
+    return task;
+  }
+
+  private OperationType determineOperationType() {
+    // TODO: determine the operation type based on the site type and kind
+    return OperationType.HARVEST;
+  }
+
+  private void queueTaskAndPutLastModifiedInSet(TaskInfo task, Set<Instant> modifiedTimes) {
+    try {
+      siteQueue.put(task);
+    } catch (InterruptedException e) {
+      modifiedTimes.add(task.getLastModified());
+      throw new QueueException(
```

Comment:
❗️ I would suggest you let that exception bubble out as opposed to convert it into an error. This is an indication that you are stopping this service based on the design. Let it bubble out, and above, handle it by updating your index if needed, persist it, and then bubble it out again (or reset the interrupted flag for the thread) before exiting the `query()` method